Create Blog “a-note-on-composing-components-with-react” (#27)
mfrachet/acodingdance: blog/a-note-on-composing-components-with-react.md

---
path: composing-components
date: 2019-05-16T18:47:58.055Z
title: A note on composing components with React
tags:
  - react
  - javascript
---
We often hear about composition when talking about frontend frameworks. Composition is the ability to create complex UIs by assembling component in as less uncoupled manner as possible.

> [Ryan Florence has made a video some years ago](https://www.youtube.com/watch?v=hEGg-3pIHlE) concerning compound components and the benefits of having well defined and composable APIs.

Author: I'll check this 😊

![Another google material card from their website](/assets/mio-design_assets_0b6xusjjsulxcutqtcvl0wurpww8_cards-dividers-2.png "Another google material card from their website")

Keeping the actual naive approach, let's imagine that we have modified the card component to achieve this:

```jsx
<Card
  avatar="https://some/path.jpg"
  title="My title"
  subtitle="My subtitle"
  image="https://some-other/path.jpg"
  description="Super funky description"
  //actionOne={() => alert('Do job')}
  //actionSecond={() => alert('Do another job')}
  //actionLike={() => alert('Do action like')}
  //actionLove={() => alert('Do action love')}
  // new implementation
  footerTitle="Footer title"
  footerSchedules={['5pm', '7am', '2pm']}
  footerAction={() => alert('Do footer stuff')}
/>
```

So as you've probably mentioned, each time we want our card to have a different look, we modify its implementation. Now let's imagine that you are the owner and maintainer of such a library. Each time people want to have a different organisation on the Card, you are supposed to modify your component, publish it to npm and manage the versions.

> It's a lot of work.

Author: Exactly, it's mentioned below 😊

## What are we doing wrong?

So here's the problem. For now, we are simply modifying the component we built in a vertical way: by adding a lots of new props on the component, and making a lot of conditional comparisons to determine the look of the card.

> There's another problem. Material Design Cards are not defined specifically. I mean, it's not supposed to provide a way to create a `PrimaryCard`, `TinyCard` and so forth. With the Material Design Card specs, you're supposed to be able to create a lots of different types of Cards without that much pain.

Author: I find this concept really hard to explain. The spec explains how we can create Cards, I mean different cards. There are no definition of a "standard" card. It's just a way to define some element that compose together.

What we have done until now actually breaks a part of the [Opened Close principle](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle). This principle explains that a code should be closed to modifications and opened to extensions (aka inheritance or composition).

> The idea is to avoid creating one single "thing" with a lot of complexe things inside, and to separate the concerns by creating multiple files.

Reviewer: maybe "entities" instead of "files"

## Material design cards using composition

> We have now a bit more of experience dealing with React component and we know that what we are doing is not sufficiently good for now. It's okay, and we'll see how we can improve using the composable nature of React.

Reviewer: We have **more** experience

Reviewer: with React component**s**

Reviewer: we know ~that~ what we are doing

Reviewer: "improve the usability"
"improve our Card component"
?

(there must be something after "improve")

> First, let's point that a piece of UI doesn't necessarily means one component. It's not always a one for one matching. I invite you reading [Brad Frost Atomic Design](http://bradfrost.com/blog/post/atomic-web-design/) if you want more information on of to build complex UIs in a more abstract way.

Reviewer: Enough "let's" 😄 

~let's point~ 👉 It's worth noticing

Reviewer: I invite you **to read**

Reviewer: "on of to build" 🤔 

👉 "about building"

> _The design we're going to implement is probably the best one but it gives a good idea of the impact of composition in React._

Reviewer: "is**n't** probably the best one" I guess

Reviewer: impact 👉 benefit?

> Let's try to split the Card component in a composable manner so that we don't have to modify its implementation when we want to create new piece of UI:

Reviewer: No "Let's" 😉

Reviewer: so ~that~ we ~d~**w**on't have to modify its implementation ~when~**if** we want to create new piece of UI:

```jsx
<Card>
  <CardHeader title="My title" subtitle="subtitle" avatar={pathToImage} />
  <CardMedia source={pathToImage} />
  <CardContent>Some content with descriptions</CardContent>
  <CardActions>
    <IconButton name="favorite" onClick={handleFav} />
    <IconButton name="star" onClick={handleStar} />
  </CardActions>
</Card>
```

Using this approach, we're able to build any type of cards:

> 

Reviewer: Why the triple carriage return from this point to the end of the post?

Author: CMS 😅

![Multiple material cards](/assets/cards003.png "Multiple material cards")

## The smell I rely on

> So this is a way to rely on composability to avoid getting complicated and unmaintainable components in React.

Reviewer: ~rely on~ 👉 legerage

Reviewer: avoid ~getting~ complicated and unmaintainable components in React

> In the future, if you have some components that start to grow, here are my "tricks" or at least the code smells (it's opinionated and depends on the cases) that indicates that I have to split my component on something a bit more composable:

Reviewer: on the **use-**cases

Reviewer: that indicate~s that~

* \- If a component has more than 5-6 props

> * \- If a component is more than 100 LoC

Reviewer: is 👉 contains

Reviewer: LoC 👉 lines of code

(the acronym may not be clear to anybody)

> * \- Can I extract this component in isolation to play in another app?

Reviewer: Can I extract this component **and reuse it** in another app?

> * \- Have I to tweak (or can I tweak) the props of the underlying components? (See this issue for more details [rn-placeholder#38](https://github.com/mfrachet/rn-placeholder/issues/38))

Reviewer: **Do I have**

> * \- Is it easy to move an underlying component somewhere else? 

Reviewer: I would rephrase this like the previous bullet and locate it right below:

>Can I extract this component and reuse it somewhere else?
>Can I extract any underlying component and reuse it somewhere else?

> But be careful! You have to find a balance between something fully composable but also easy to use. Remember that abstracting is not always a good thing.

Reviewer: **the right** balance

Reviewer: tell why

Maybe by ending the sentence with "if it makes your code more verbose and less readable"

---
path: composing-components
date: 2019-05-16T18:47:58.055Z
title: A note on composing components with React
tags:
  - react
  - javascript
---

> We often hear about composition when talking about frontend frameworks. Composition is the ability to create complex UIs by assembling components in the less uncoupled manner as possible.

Reviewer 2: I don't understand this `in the less uncoupled manner as possible`. You mean `less coupled manner`, right ? Otherwise I would be glad to have further explanations 👍

Author: Yep 😅 THanks for spotting this!

[Ryan Florence has made a video some years ago](https://www.youtube.com/watch?v=hEGg-3pIHlE) concerning compound components and the benefits of having well-defined and composable APIs.

<iframe width="100%" height="400" src="https://www.youtube.com/embed/hEGg-3pIHlE" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

> I'm a software developer and I make mistakes. And some years ago, I've released [rn-placeholder](https://github.com/mfrachet/rn-placeholder) using an API that was not extensible enough because of lack of composability.

Reviewer 2: ```suggestion
I'm a software developer and I make mistakes. Some years ago, I've released [rn-placeholder](https://github.com/mfrachet/rn-placeholder) using an API that was not extensible enough because of lack of composability.
```

And is unnecessary imho 😄

## Implementing using a naive approach

Let's take the example of the [Material Design Cards](https://material.io/design/components/cards.html). A "Card" in the Material Design specification is a way to present heterogeneous content using a homogeneous UI representation.

![A card definition from the Material Design Card specification](/assets/mio-design_assets_1ezntdj8h1j0bfkbl23lyzewjjvmzy_uv_cards-elements-2b.png "A card definition from the Material Design Card specification")

As with a naive approach, here's the way I would have defined my API at first:

```jsx
<Card
  avatar="https://some/path.jpg"
  title="My title"
  subtitle="My subtitle"
  image="https://some-other/path.jpg"
  description="Super funky description"
  actionOne={() => console.log('Do job')}
  actionSecond={() => console.log('Do another job')}
  actionLike={() => console.log('Do action like')}
  actionLove={() => console.log('Do action love')}
/>
```

And to be honest, the previous code snippet is a valid code and it can work in real world applications.

> The problem comes when you start customising the Card to handle its variations. Because as I mentioned above, Card in the Material Design specification is a way to represent heterogeneous information using some homogeneous styles.

Reviewer 2: ```suggestion
The problem comes when you start customizing the Card to handle its variations. Because as I mentioned above, Card in the Material Design specification is a way to represent heterogeneous information using some homogeneous styles.
```

> The problem comes when you start customising the Card to handle its variations. Because as I mentioned above, Card in the Material Design specification is a way to represent heterogeneous information using some homogeneous styles.

Reviewer 2: I feel like "As I mentionned" is enough you don't need "Because" 😄

Let's imagine that we want to modify the actual card component to look like:

![Another Material Card from their website](/assets/mio-design_assets_0b6xusjjsulxcutqtcvl0wurpww8_cards-dividers-2.png "Another Material Card from their website")

Keeping the actual naive approach, let's imagine that we have modified the card component to achieve this:

```jsx
<Card
  avatar="https://some/path.jpg"
  title="My title"
  subtitle="My subtitle"
  image="https://some-other/path.jpg"
  description="Super funky description"
  //actionOne={() => console.log('Do job')}
  //actionSecond={() => console.log('Do another job')}
  //actionLike={() => console.log('Do action like')}
  //actionLove={() => console.log('Do action love')}
  // new implementation
  footerTitle="Footer title"
  footerSchedules={['5pm', '7am', '2pm']}
  footerAction={() => console.log('Do footer stuff')}
/>
```

> So as you've probably noticed, each time we want our card to have a different look, we modify its implementation. Now imagine yourself as an owner or maintainer of such a library. Each time people want to have a different organisation on the Card, you are supposed to modify your component, publish it to npm and manage the versions.

Reviewer 2: ```suggestion
As you've probably noticed, each time we want our card to have a different look, we modify its implementation. Now imagine yourself as an owner or maintainer of such a library. Each time people want to have a different organisation on the Card, you are supposed to modify your component, publish it to npm and manage the versions.
```

So is more a spoken word, I guess ?

_Edit: You use it a few times later (in case you want to remove it 😉)_

It's a lot of work.

## What are we doing wrong?

> So here's the problem. For now, we are simply modifying the component we built in a vertical way: by adding a lot of new props on the component, and making a lot of conditional comparisons to determine the look and feel of the card.

Reviewer 2: ```suggestion
So here's the problem. For now, we are simply modifying the component we built in a vertical way by adding a lot of new props on the component and making a lot of conditional comparisons to determine the look and feel of the card.
```

There's another problem. Material Design Cards are not defined specifically. I mean, it's not supposed to provide a way to create a `PrimaryCard`, `TinyCard` and so forth. With the Material Design Card specs, you're supposed to be able to create a lots of different types of Cards without that much pain.

What we have done until now actually breaks a part of the [Opened Closed principle](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle). This principle tells that a piece of code should be closed to modifications and opened to extensions (aka inheritance or composition).

The idea is to avoid creating one single "thing" with a lot of complex things inside, but separate the concerns by creating multiple entities.

## Material design cards using composition

> We have more of experience dealing with React components and we know what we are doing is not sufficiently good for now. It's okay, and we'll see how we can improve the Card component API using the composable nature of React.

Reviewer 2: ```suggestion
We have more of experience dealing with React components and we know what we are doing is not good enough yet. It's okay, we'll see how we can improve the Card component API using the composable nature of React.
```

Reviewer 2: I found a [nice article](https://www.grammarly.com/blog/comma-before-and/) about comma before "and" 😄

> First, it's worth noticing that a piece of UI doesn't necessarily means one component. It's not always a one for one matching. I invite you to read [Brad Frost Atomic Design](http://bradfrost.com/blog/post/atomic-web-design/) if you want more information about building complex UIs in a more abstract way.

Reviewer 2: ```suggestion
First, it's worth noticing that a piece of UI doesn't necessarily means one component. It's not always a one to one matching. I invite you to read [Brad Frost Atomic Design](http://bradfrost.com/blog/post/atomic-web-design/) if you want more informations about building complex UIs in a more abstract way.
```

> _The design we're going to implement isn't probably the best one but it gives a good idea of the benefit of composition in React._

Reviewer 2: ```suggestion
_The design we're going to implement isn't probably the best one but it gives a good idea of the benefits of composition in React._
```

I've splitted the Card component in a composable manner so we won't have to modify its implementation if we want to create new piece of UI:

```jsx
<Card>
  <CardHeader title="My title" subtitle="subtitle" avatar={pathToImage} />
  <CardMedia source={pathToImage} />
  <CardContent>Some content with descriptions</CardContent>
  <CardActions>
    <IconButton name="favorite" onClick={handleFav} />
    <IconButton name="star" onClick={handleStar} />
  </CardActions>
</Card>
```

Using this approach, we're able to build any type of cards:

![Multiple material cards](/assets/cards003.png "Multiple material cards")

## The smell I rely on

> So this is a way to leverage composability to avoid complicated and unmaintainable components in React.

Reviewer 2: ```suggestion
So this is a way to leverage composability to avoid complex and unmaintainable components in React.
```

In the future, if you have some components that start to grow, here are my "tricks" or at least the code smells (it's opinionated and depends on the use-cases) that indicates I have to split my component on something a bit more composable:

* If a component has more than 5-6 props
* If a component contains more than 100 lines of code
* Can I extract this component and reuse it in another app?
* Do I have to tweak (or can I tweak) the props of the underlying components? (See this issue for more details [rn-placeholder#38](https://github.com/mfrachet/rn-placeholder/issues/38))
* Can I move the underlying component somewhere else in the parent component (top or bottom)? 

> But be careful! You have to find the right balance between something fully composable but also easy to use. Remember that abstracting is not always a good thing.

Reviewer 2: ```suggestion
But be careful! You have to find the right balance between something fully composable and easy to use. Remember that abstracting is not always a good thing.
```

> If you find yourself writing tons of line to create a component you know you'll use only one time, it's probably not necessary to find a good abstraction. However if it's a component that will leave in multiple places and that can probably move in terms of shape, you maybe have to think twice while implementing.

Reviewer 2: ```suggestion
If you find yourself writing tons of lines to create a component you know you'll use it only one time, it's probably not necessary to find a good abstraction. However if it's a component that you will use in multiple places and that can probably move in terms of shape, you maybe have to think twice while implementing.
```

> And don't forget that with the new composable API, we can create a component that just looks like the old one!

Reviewer 2: ```suggestion
And don't forget with the new composable API, we can create a component that just looks like the old one!
```
